docs/design.md

```diff
@@ -0,0 +1,128 @@
+
+# Clients
+A default client,
+[SimpleClient](../bftengine/include/bftengine/SimpleClient.hpp), is included
+with concord-bft. It is just one manner of implementing a client, and is
+implemented in C++ utilizing the `ICommunication` and `IReceiver` interfaces
+below. Additional clients can be implemented by utilizing the ClientMessages
+described in the next subsection. In fact, we have a [client in
+Python](../util/pyclient/bft_client.py) that we use for testing that does just
+that.
+
+## Client Messages
+Client request and reply messages are prefixed with a `ClientRequestHeader` or
+`ClientReplyHeader` respectively. Each header, like all other messages in
+concord-bft is described as a packed C struct, meaning that there are no padding
+bytes between struct fields. This is accomplished via enclosing the structure
+definitions between `#pragma pack(push,1)` and `#pragma pack(pop)` lines.
+Packing ensures that the size of the structure is the sum of the size of the
+fields, and allows them to be sent over the wire and interpreted
+deterministically, given both machines have the same endianess.
+
+The primary benefit of this approach is that it is optimal in time and space
+requirements. Structs are exactly as large as they need to be, and there is no
+need to serialize and deserialize them, as they are operated on in memory
+identically to how they are sent on the wire. It is also very easy to
+implement and understand.
+
+The drawbacks are:
+ * Endianness is not taken into consideration, so a struct sent from a big
+   endian machine will be incorrectly interpreted on a little endian machine,
+   and vice versa.
+ * Messages field access may be slow, as it may not be on an aligned boundary.
+ * Messages are fixed. New fields cannot be added to messages and remain
+   backwards compatible.
+
+## Communication
+
+The `ICommunication` interface is used by the bftengine code to send
+asynchronous messages across the network. It is implemented for different
+transport layers such as UDP, TCP, and TLS in
+[bftengine/src/communication](../bftengine/src/communication).
+
+The `setReceiver` method is called by the bftengine code to let implementers of
+the `ICommunication` interface know where received messages should be handled.
+Handlers for received messages implement the `IReceiver` interface. The two sole
+implementers of `IReceiver` are the [ReplicaImp](../bftengine/src/bftengine/ReplicaImp.hpp)
+and the [SimpleClient](../bftengine/src/bftengine/SimpleClient.cpp)
+
+## State transfer
+
+State transfer refers to the mechanism via which out of date replicas can
+retrieve already committed data efficiently and bring themselves in sync with
+the rest of the cluster. State transfer is done outside of the ordering and
+replication protocol performed by SBFT, and can often be achieved by data transfer
+between only two replicas.
+
+As state transfer is outside of the core SMR protocol, it may be implemented in
```

Comment:
I am not sure saying its outside the core SMR is correct. Maybe say outside the Byzantine Agreement protocol?  Also I think we need to say that ST must be implemented as a function of the instantiation of the specific state machine implementation (the term application may be confusing in this context). This is related to my comment above about the discussion of Concord implementing the BASE approach of having an abstract state machine interface.

Reply:
Updated as you described. Thanks!